### PR TITLE
Balder/produce summary features if using dispatch.summaries

### PR DESCRIPTION
**Pull request by baldersheim**
@bratseth PR java code.
@havardpe PR The rest .... Probably easier to review commit by commit. The semantic changes are in fbd92b4, 5905c45, 00f7a50 and 5d8d28c

### CODE REVIEW

File: searchlib/src/vespa/searchlib/engine/propertiesmap.h [hunk 1]
```diff
@@ -111,21 +110,22 @@ class PropertiesMap
      *
      * @return cache properties
      **/
-    const search::fef::Properties &cacheProperties() const {
+    const Props &cacheProperties() const {
         return lookup(MapNames::CACHES);
     }
+    Props &cacheProperties() {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/engine/propertiesmap.h]: This should not be overloaded since it alters inspection behavior based on object constness.

**Review by havardpe — CHANGES_REQUESTED** [review]
please consider my comment.
Also; will there be some Java code added to handle the case where you want summaryfeatures but have session cache disabled? (error, auto-enable,...?)

**Comment by bratseth** [comment]
Container changes : +1 on Håvard's comment (error probably), LGTM otherwise

File: searchlib/src/vespa/searchlib/engine/propertiesmap.h (hunk 1 above)
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/engine/propertiesmap.h]: Fixed

**Comment by baldersheim** [comment]
All comments fixed, merging.

**Comment by havardpe** [comment]
👍

**Comment by bratseth** [comment]
But did you do something about the summaryfeatures + no cache combination?

**Comment by baldersheim** [comment]
No, you just do not get them. I consider that an orthogonal issue.

**Comment by baldersheim** [comment]
After some pondering I concluded that either adding magic or detecting inconsistency and produce an error either adding extra cost or very complicated.

1 - The simple solution would be to enable querycache if dispatch.summaries were set. That would ensure correct behaviour. But it would also add cpu/memory cost for the cases where it is not needed. And those are also the cases dispatch.summaries were initially intended for.

2 - The perfect solution would need to examine both rank-profile for presence of summary features, and examine summary class for presence of rewriters needing access to either query or location. I am not sure how much of this information is present/easily accessible in the qrs.

**Comment by baldersheim** [comment]
After poking around I found FastSearcher.summaryNeedsQuery that might already be doing the hard stuff. So then I will produce an error during search. This will force them to use summary/rankprofile that are consistent with query parameters.

**Pull request merged**